### ISSUE TITLE
Very long compilation time on big projects

### ISSUE THREAD

**Issue by AndreyZM** [bridgedotnet/Bridge#1902]
I have note that Bridge take a lot of time to compile our project(about ~11 minutes). And i note that compilation without Bridge take about 10-20sec. 

```
2016-09-30T01:28:41:870     Info        Sorting references done
2016-09-30T01:28:41:870     Info        Before emitting...
2016-09-30T01:28:41:870     Info        Before emitting done
2016-09-30T01:28:41:871     Info        Emitting...
2016-09-30T01:28:41:871     Trace       Emitting block Bridge.Translator.EmitBlock
//8 min. later
2016-09-30T01:28:49:761     Info        Transforming outputs...
2016-09-30T01:28:49:762     Trace       Wrapping to modules...
2016-09-30T01:28:49:762     Trace       Wrapping to modules done
2016-09-30T01:28:49:762     Trace       Combining outputs...
```

**Comment by LeonidVeriga** [comment]
Can you check compilation time with logging turned off:
```
"logging:" {
    "level": "None"
}
```
How many classes does the project contain?

**Comment by kcudnik** [comment]
10-20 sec compilation using visual studio compiler? How many lines/kB does your project have?

**Comment by AndreyZM** [comment]
full project have about 3k source files...

**Comment by kcudnik** [comment]
Wow, you can probably split that into libraries instead of compiling everything in one shot every time, should help.

**Comment by AndreyZM** [comment]
Its splited. But compilation root library cause full rebuild and its take about 15min

**Comment by kcudnik** [comment]
I dont know your project settings, but full rebuild should be only done when you ask for that explicitly, when you reference other bridge dlls with embedded js it should be a lot faster

**Comment by AndreyZM** [comment]
Full rebuild happen when we change the root library, because all projects in solution depends on it.

**Comment by AndreyZM** [comment]
I there way to profile Bridge? May be i can find a bottle neck

**Comment by vladsch** [comment]
If you want profile compilation process then clone Bridge repository, add sandbox project to Bridge solution and run translation from sandbox (also it allows to debug Bridge)

```cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using Bridge.Contract;
using Bridge.Translator;
using Bridge.Translator.Logging;

namespace Bridge.Sandbox
{
    class Program
    {
        static void Main(string[] args)
        {
            var logger = new Logger("Bridge.Builder.Console", true, new ConsoleLoggerWriter(), new FileLoggerWriter());
            logger.LoggerLevel = LoggerLevel.Error;
            // PLEASE ENSURE THAT Bridge.dll file is available in the output folder (bin/Debug/)

            var root = @"C:\projects\Bridge\v1\";
            var projectLocation = root + @"Sandbox\TestLibrary\Bridge.TestLibrary.csproj ";            
            var clrLocation = root + @"Bridge\Bridge\bin\Debug\Bridge.dll";
            var outputLocation = Path.ChangeExtension(projectLocation, "js");

            var bridgeOptions = new Bridge.Translator.BridgeOptions()
            {
                ProjectLocation = projectLocation,
                OutputLocation = outputLocation,
                BridgeLocation = clrLocation,
                Rebuild = false,
                ExtractCore = true,
                Configuration = null,
                Source = null,
                Folder = null,
                Recursive = false,
                Lib = null,
                DefinitionConstants = null,
                Help = false,
                NoTimeStamp = null,
                FromTask = false,
                Name = "Sandbox"
            };

            TimeSpan begin = Process.GetCurrentProcess().TotalProcessorTime;

            var processor = new TranslatorProcessor(bridgeOptions, logger);
            
            var result = processor.PreProcess();

            if (result != null)
            {
                return;
            }

            try
            {
                processor.Process();
                processor.PostProcess();
                TimeSpan end = Process.GetCurrentProcess().TotalProcessorTime;
                Console.WriteLine("Compilation time: " + (end - begin).TotalSeconds + " ms.");

                Console.WriteLine(processor.Translator.GetCode());
                Console.ReadLine();
            }
            catch (Exception e)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Error: {0}", e.Message);
                Console.ResetColor();
                Console.ReadLine();
            }
        }
    }
}

```

**Comment by LeonidVeriga** [comment]
@AndreyZM, can you check https://github.com/bridgedotnet/Bridge/issues/1902#issuecomment-250585489

**Comment by AndreyZM** [comment]
In the log above you can see that betwen "emmiting block" and "transforming outputs"

**Comment by geoffreymcgill** [comment]
@AndreyZM - Was this Issue closed accidentally?

**Comment by AndreyZM** [comment]
In the log above you can see that 8min betwen "emmiting block" and "transforming outputs", and there no any message betwen.

**Comment by AndreyZM** [comment]
Hi again. We can provide a stripped project that took 5 minutes to compile with a Bridge. May be it can help to profile and optimize. Just give me a email to mail to.

**Comment by geoffreymcgill** [comment]
Hi @AndreyZM - Can you create a GitHub repo for the sample project? That would be much better and allow for efficient collaboration.

**Comment by AndreyZM** [comment]
I just don't want to put it into public domain.

**Comment by AndreyZM** [comment]
Here is
https://github.com/Shkarin/TestBuildBridge

**Comment by geoffreymcgill** [comment]
HI @AndreyZM - Thanks for the demo project. We were able to make some improvements. With our revisions, your project is now compiling 300% faster for us.

More improvements will come in stages. These recent changes will be included in 15.4, and others are planned for 16.0 (or beyond).

At least your project is compiling in a much more reasonable time now, and we're still investigating a few other revisions that will hopefully further improve those numbers for the 15.4 release. 

More information will be posted once the changes have hit the **master** branch.

**Comment by LeonidVeriga** [comment]
Compilation time before and after

| | Before | After |
| --- | --- | --- |
| [TestProject](https://github.com/Shkarin/TestBuildBridge) | 3min | 33sec |
| Batch1 | 37sec | 22.54sec |
| Batch2 | 5.45sec | 5.7sec |
| Batch3 | 14.98sec | 10.75sec |
| Batch4 | 25.1sec | 17.85sec |
